<a href="https://colab.research.google.com/github/amatchison/CS4410/blob/main/Homework7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [73]:
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D

# **Exercise 16.1**

In [66]:
#create neural network
cnn = Sequential()

In [67]:
#add layers
cnn.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

cnn.add(Flatten())

cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))

In [68]:
#compile model
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [69]:
#load fasion_mnist
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

#reshape data
X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))

#normalize
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

#one-hot encoding
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

In [70]:
#train model
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 [==============================] - 109s 128ms/step - loss: 0.4646 - accuracy: 0.8338 - val_loss: 0.3271 - val_accuracy: 0.8835
Epoch 2/5
844/844 [==============================] - 106s 126ms/step - loss: 0.3020 - accuracy: 0.8908 - val_loss: 0.2934 - val_accuracy: 0.8915
Epoch 3/5
844/844 [==============================] - 109s 129ms/step - loss: 0.2551 - accuracy: 0.9064 - val_loss: 0.2597 - val_accuracy: 0.9042
Epoch 4/5
844/844 [==============================] - 109s 129ms/step - loss: 0.2222 - accuracy: 0.9179 - val_loss: 0.2576 - val_accuracy: 0.9027
Epoch 5/5
844/844 [==============================] - 109s 129ms/step - loss: 0.1947 - accuracy: 0.9269 - val_loss: 0.2509 - val_accuracy: 0.9070


In [71]:
#evaluate model
loss, accuracy = cnn.evaluate(X_test, y_test)

313/313 [==============================] - 7s 21ms/step - loss: 0.2693 - accuracy: 0.9017


In [72]:
#load mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

#reshape data
X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))

#normalize
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

#one-hot encoding
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

11490434/11490434 [==============================] - 0s 0us/step


In [74]:
#train model
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 [==============================] - 104s 123ms/step - loss: 0.1241 - accuracy: 0.9647 - val_loss: 0.0373 - val_accuracy: 0.9897
Epoch 2/5
844/844 [==============================] - 105s 124ms/step - loss: 0.0329 - accuracy: 0.9902 - val_loss: 0.0353 - val_accuracy: 0.9900
Epoch 3/5
844/844 [==============================] - 105s 124ms/step - loss: 0.0207 - accuracy: 0.9933 - val_loss: 0.0363 - val_accuracy: 0.9908
Epoch 4/5
844/844 [==============================] - 105s 124ms/step - loss: 0.0156 - accuracy: 0.9951 - val_loss: 0.0383 - val_accuracy: 0.9907
Epoch 5/5
844/844 [==============================] - 104s 123ms/step - loss: 0.0109 - accuracy: 0.9964 - val_loss: 0.0364 - val_accuracy: 0.9905


In [75]:
#evaluate model
loss, accuracy = cnn.evaluate(X_test, y_test)

313/313 [==============================] - 7s 21ms/step - loss: 0.0332 - accuracy: 0.9916




---


# How well does the model perform on Fashion-MNIST compared to MNIST?
The model performs slightly better on MNIST than on Fashion-MNIST.

Fashion-MNIST had a loss of 0.27 and an accuracy of 0.90. MNIST had 0.03 loss and 0.99 accuracy.


# How do the training times compare?
Both had similar training times of around 105 - 110s per epoch.

# **Exercise 16.4**

In [76]:
#create neural network 2
cnn2 = Sequential()

#add layers
cnn2.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn2.add(MaxPooling2D(pool_size=(2, 2)))

cnn2.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn2.add(MaxPooling2D(pool_size=(2, 2)))

cnn2.add(Flatten())
#REMOVED 1ST DENSE LAYER HERE
cnn2.add(Dense(units=10, activation='softmax'))

In [77]:
#compile model
cnn2.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [78]:
#train model
cnn2.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 [==============================] - 107s 126ms/step - loss: 0.1708 - accuracy: 0.9487 - val_loss: 0.0502 - val_accuracy: 0.9867
Epoch 2/5
844/844 [==============================] - 105s 125ms/step - loss: 0.0507 - accuracy: 0.9843 - val_loss: 0.0388 - val_accuracy: 0.9890
Epoch 3/5
844/844 [==============================] - 103s 122ms/step - loss: 0.0353 - accuracy: 0.9888 - val_loss: 0.0392 - val_accuracy: 0.9898
Epoch 4/5
844/844 [==============================] - 103s 122ms/step - loss: 0.0279 - accuracy: 0.9909 - val_loss: 0.0313 - val_accuracy: 0.9907
Epoch 5/5
844/844 [==============================] - 105s 125ms/step - loss: 0.0211 - accuracy: 0.9930 - val_loss: 0.0345 - val_accuracy: 0.9902


In [80]:
#evaluate model
loss, accuracy = cnn2.evaluate(X_test, y_test)

313/313 [==============================] - 5s 17ms/step - loss: 0.0330 - accuracy: 0.9890


# How does this change prediction accuracy?

Prediction accuracy was nearly unchanged, dropping from 0.99 to 0.98

Speed shows no noteworthy difference, remaining around 105s per epoch.

In [81]:
#create neural network 3
cnn3 = Sequential()

#add layers
cnn3.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn3.add(MaxPooling2D(pool_size=(2, 2)))

cnn3.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn3.add(MaxPooling2D(pool_size=(2, 2)))

cnn3.add(Flatten())
cnn3.add(Dense(units=4096, activation='relu'))
cnn3.add(Dense(units=128, activation='relu'))
cnn3.add(Dense(units=10, activation='softmax'))

In [82]:
#compile model
cnn3.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [83]:
#train model
cnn3.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 [==============================] - 371s 438ms/step - loss: 0.1254 - accuracy: 0.9614 - val_loss: 0.0541 - val_accuracy: 0.9838
Epoch 2/5
844/844 [==============================] - 389s 461ms/step - loss: 0.0402 - accuracy: 0.9873 - val_loss: 0.0388 - val_accuracy: 0.9882
Epoch 3/5
844/844 [==============================] - 381s 452ms/step - loss: 0.0272 - accuracy: 0.9915 - val_loss: 0.0338 - val_accuracy: 0.9900
Epoch 4/5
844/844 [==============================] - 392s 465ms/step - loss: 0.0180 - accuracy: 0.9941 - val_loss: 0.0359 - val_accuracy: 0.9905
Epoch 5/5
844/844 [==============================] - 400s 474ms/step - loss: 0.0167 - accuracy: 0.9948 - val_loss: 0.0372 - val_accuracy: 0.9908


In [84]:
#evaluate model
loss, accuracy = cnn3.evaluate(X_test, y_test)

313/313 [==============================] - 15s 47ms/step - loss: 0.0363 - accuracy: 0.9895


# How does this change prediction accuracy?

Prediction accuracy was nearly unchanged again, dropping from 0.99 to 0.989.

For some reason, speed dropped drastically, taking nearly 4 times longer to train.